File: data_structures/queues/queue_by_list.cpp
#include <deque>
#include <iostream>
#include <stdexcept>

// Queue implementation using std::deque to support rotation efficiently
template <typename T> class Queue {
private:
  std::deque<T> entries;

public:
  void put(T item) { entries.push_back(item); }

  T get() {
    if (entries.empty()) {
      throw std::out_of_range("Queue is empty");
    }
    T item = entries.front();
    entries.pop_front();
    return item;
  }

  T get_front() const {
    if (entries.empty()) {
      throw std::out_of_range("Queue is empty");
    }
    return entries.front();
  }

  void rotate(int rotation) {
    if (entries.empty())
      return;
    for (int i = 0; i < rotation; ++i) {
      T item = entries.front();
      entries.pop_front();
      entries.push_back(item);
    }
  }

  bool isEmpty() const { return entries.empty(); }

  int size() const { return entries.size(); }
};

int main() {
  Queue<int> q;
  q.put(10);
  q.put(20);
  q.put(30);

  std::cout << "Front: " << q.get_front() << std::endl; // 10
  q.rotate(1);
  std::cout << "Front after rotation: " << q.get_front() << std::endl; // 20

  std::cout << "Get: " << q.get() << std::endl;   // 20
  std::cout << "Size: " << q.size() << std::endl; // 2

  return 0;
}

File: data_structures/queues/queue_by_list.rs
use std::collections::VecDeque;

#[derive(Debug)]
pub struct Queue<T> {
    entries: VecDeque<T>,
}

impl<T> Queue<T> {
    pub fn new() -> Self {
        Queue {
            entries: VecDeque::new(),
        }
    }

    pub fn put(&mut self, item: T) {
        self.entries.push_back(item);
    }

    pub fn get(&mut self) -> Option<T> {
        self.entries.pop_front()
    }

    pub fn get_front(&self) -> Option<&T> {
        self.entries.front()
    }

    pub fn rotate(&mut self, rotation: usize) {
        if self.entries.is_empty() {
            return;
        }
        for _ in 0..rotation {
            if let Some(item) = self.entries.pop_front() {
                self.entries.push_back(item);
            }
        }
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}

fn main() {
    let mut q = Queue::new();
    q.put(10);
    q.put(20);
    q.put(30);
    
    println!("Front: {:?}", q.get_front());
    q.rotate(1);
    println!("Front after rotation: {:?}", q.get_front());
    println!("Get: {:?}", q.get());
}
